Failure when generating empty union

## What happened?

When we have a union type with no sub-types:
```
MyEmptyUnion:
  union: {}
```

The following class is generated:
```
export type IMyEmptyUnion;

export interface IMyEmptyUnionVisitor<T> {
    'unknown': (obj: IMyEmptyUnion) => T;
}

function visit<T>(obj: IMyEmptyUnion, visitor: IMyEmptyUnionVisitor<T>): T {
    return visitor.unknown(obj);
}

export const IMyEmptyUnion = {
    visit: visit
};
```

And `compileTypeScript` step fails with:
```
myEmptyUnion.ts(1,26): error TS4081: Exported type alias 'IMyEmptyUnion' has or is using private name ''.
myEmptyUnion.ts(1,26): error TS1005: '=' expected.
```

## What did you want to happen?

I'd expect a compilable empty union class, or I'd expect Conjure spec layer to disallow empty unions.

## More details
This bug was introduced with the version [5.6.0](https://github.com/palantir/conjure-typescript/releases/tag/5.6.0), looking at the diff https://github.com/palantir/conjure-typescript/compare/5.5.0...5.6.0, I suspect https://github.com/palantir/conjure-typescript/pull/247 might have caused it.

--- a/changelog/@unreleased/pr-302.v2.yml
+++ b/changelog/@unreleased/pr-302.v2.yml
@@ -1,0 +1,5 @@
+type: fix
+fix:
+  description: Fix the bug where `compileTypeScript` task fails for empty unions
+  links:
+  - https://github.com/palantir/conjure-typescript/pull/302

--- a/src/commands/generate/__tests__/resources/definitions/example-types.yml
+++ b/src/commands/generate/__tests__/resources/definitions/example-types.yml
@@ -163,6 +163,8 @@ types:
           if: integer # some 'bad' member names!
           new: integer
           interface: integer
+      EmptyUnionTypeExample:
+        union: {}
       EmptyObjectExample:
         fields: {}
       AliasAsMapKeyExample:

--- a/src/commands/generate/__tests__/resources/flavored-test-cases/example-types/product/emptyUnionTypeExample.ts
+++ b/src/commands/generate/__tests__/resources/flavored-test-cases/example-types/product/emptyUnionTypeExample.ts
@@ -1,4 +1,4 @@
-export type IEmptyUnionTypeExample = any;
+export type IEmptyUnionTypeExample = unknown;
 
 export interface IEmptyUnionTypeExampleVisitor<T> {
     'unknown': (obj: IEmptyUnionTypeExample) => T;

--- a/src/commands/generate/__tests__/resources/flavored-test-cases/example-types/product/emptyUnionTypeExample.ts
+++ b/src/commands/generate/__tests__/resources/flavored-test-cases/example-types/product/emptyUnionTypeExample.ts
@@ -1,0 +1,13 @@
+export type IEmptyUnionTypeExample = any;
+
+export interface IEmptyUnionTypeExampleVisitor<T> {
+    'unknown': (obj: IEmptyUnionTypeExample) => T;
+}
+
+function visit<T>(obj: IEmptyUnionTypeExample, visitor: IEmptyUnionTypeExampleVisitor<T>): T {
+    return visitor.unknown(obj);
+}
+
+export const IEmptyUnionTypeExample = {
+    visit: visit
+};

--- a/src/commands/generate/__tests__/resources/readonly-test-cases/example-types/product/emptyUnionTypeExample.ts
+++ b/src/commands/generate/__tests__/resources/readonly-test-cases/example-types/product/emptyUnionTypeExample.ts
@@ -1,4 +1,4 @@
-export type IEmptyUnionTypeExample = any;
+export type IEmptyUnionTypeExample = unknown;
 
 export interface IEmptyUnionTypeExampleVisitor<T> {
     readonly 'unknown': (obj: IEmptyUnionTypeExample) => T;

--- a/src/commands/generate/__tests__/resources/readonly-test-cases/example-types/product/emptyUnionTypeExample.ts
+++ b/src/commands/generate/__tests__/resources/readonly-test-cases/example-types/product/emptyUnionTypeExample.ts
@@ -1,0 +1,13 @@
+export type IEmptyUnionTypeExample = any;
+
+export interface IEmptyUnionTypeExampleVisitor<T> {
+    readonly 'unknown': (obj: IEmptyUnionTypeExample) => T;
+}
+
+function visit<T>(obj: IEmptyUnionTypeExample, visitor: IEmptyUnionTypeExampleVisitor<T>): T {
+    return visitor.unknown(obj);
+}
+
+export const IEmptyUnionTypeExample = {
+    visit: visit
+};

--- a/src/commands/generate/__tests__/resources/test-cases/example-types/product/emptyUnionTypeExample.ts
+++ b/src/commands/generate/__tests__/resources/test-cases/example-types/product/emptyUnionTypeExample.ts
@@ -1,4 +1,4 @@
-export type IEmptyUnionTypeExample = any;
+export type IEmptyUnionTypeExample = unknown;
 
 export interface IEmptyUnionTypeExampleVisitor<T> {
     'unknown': (obj: IEmptyUnionTypeExample) => T;

--- a/src/commands/generate/__tests__/resources/test-cases/example-types/product/emptyUnionTypeExample.ts
+++ b/src/commands/generate/__tests__/resources/test-cases/example-types/product/emptyUnionTypeExample.ts
@@ -1,0 +1,13 @@
+export type IEmptyUnionTypeExample = any;
+
+export interface IEmptyUnionTypeExampleVisitor<T> {
+    'unknown': (obj: IEmptyUnionTypeExample) => T;
+}
+
+function visit<T>(obj: IEmptyUnionTypeExample, visitor: IEmptyUnionTypeExampleVisitor<T>): T {
+    return visitor.unknown(obj);
+}
+
+export const IEmptyUnionTypeExample = {
+    visit: visit
+};

--- a/src/commands/generate/typeGenerator.ts
+++ b/src/commands/generate/typeGenerator.ts
@@ -269,7 +269,7 @@ export async function generateUnion(
         name: unionTsType,
         type:
             unionSourceFileInput.memberInterfaces.length === 0
-                ? "any"
+                ? "unknown"
                 : unionSourceFileInput.memberInterfaces.map(iface => iface.name).join(" | "),
     });
 

--- a/src/commands/generate/typeGenerator.ts
+++ b/src/commands/generate/typeGenerator.ts
@@ -267,7 +267,10 @@ export async function generateUnion(
         docs: definition.docs != null ? [{ description: definition.docs }] : undefined,
         isExported: true,
         name: unionTsType,
-        type: unionSourceFileInput.memberInterfaces.map(iface => iface.name).join(" | "),
+        type:
+            unionSourceFileInput.memberInterfaces.length === 0
+                ? "any"
+                : unionSourceFileInput.memberInterfaces.map(iface => iface.name).join(" | "),
     });
 
     const visitorInterface = sourceFile.addInterface({

